### PR TITLE
Big integer as a record

### PR DESCRIPTION
**Pull request by noakesmichael**
Revise the record-based version of the Big Integer wrapping of GMP that was started by
Laura Delaney.

For the next release the record Bigint will sit alongside the class BigInt.  However, after a subsequent
PR, use of BigInt will generate a "deprecated" message that will forward to Bigint.

This PR is a solid implementation of the record Bigint, updates the current tests of BigInt to use Bigint, and adds some tests by Laura.  I have performed some limited multi-locale testing.  However
I intend to do more testing during the coming week so that this is as strong as possible for the
release.

Targeted testing on clang/darwin with and without gasnet.
Full para test on linux64 for single locale and gasnet.

### CODE REVIEW

**Comment by benharsh** [comment]
Have you had a chance to run this through chpldoc and look to see if it renders OK? We have a week until doc freeze, so I wouldn't hold this PR up over it.

**Comment by noakesmichael** [comment]
Good reminder.  I haven’t done this yet but intend to well before doc freeze.



On Sep 19, 2016, at 3:25 PM, Ben Harshbarger <notifications@github.com<mailto:notifications@github.com>> wrote:


Have you had a chance to run this through chpldoc and look to see if it renders OK? We have a week until doc freeze, so I wouldn't hold this PR up over it.

File: modules/standard/BigInteger.chpl [hunk 1]
```diff
@@ -2788,7 +2790,7 @@ module BigInteger {
     const sb_ = starting_bit.safeCast(c_ulong);
     var   ret: c_ulong;
 
-    if _local || this.locale_id == chpl_nodeID {
+    if _local || this.locaelId == chpl_nodeID {
```
Comment (benharsh) [inline modules/standard/BigInteger.chpl]: typo for 'locaelId'

**Comment by benharsh** [comment]
I didn't review the entire set of changes to BigInteger thoroughly, but did so for a handful of operations to familiarize myself with the various patterns. The current approach looks reasonable to me, and seems to be consistent for the rest of the operations.

Provided that other typos are fixed up, this looks good to merge 👍

**Comment by noakesmichael** [comment]
Thanks.  Full gasnet testing caught one of these and then I global replaced
for the many cut/paste replicants.

Mike


On Sep 19, 2016, at 3:50 PM, Ben Harshbarger <notifications@github.com<mailto:notifications@github.com>> wrote:


@benharsh commented on this pull request.

________________________________

In modules/standard/BigInteger.chpl<https://github.com/chapel-lang/chapel/pull/4572#pullrequestreview-653673>:

> @@ -2788,7 +2790,7 @@ module BigInteger {
     const sb_ = starting_bit.safeCast(c_ulong);
     var   ret: c_ulong;

-    if _local || this.locale_id == chpl_nodeID {
+    if _local || this.locaelId == chpl_nodeID {


typo for 'locaelId'

**Pull request merged**